torch_xla/csrc/xla_sharding_util.cpp

```diff
@@ -382,4 +383,64 @@ std::vector<at::Tensor> ShardingUtil::ShardTensor(
   return shards;
 }
 
+void ShardingUtil::PrepareOutputShardingPropagation(
+    std::vector<XLATensorPtr>* tensors, absl::Span<const size_t> indices,
+    ComputationPtr computation,
+    std::vector<torch::lazy::BackendDataPtr>* data_placeholders,
+    std::vector<XLATensor::ShardingSpecPtr>* sharding_specs) {
+  auto computation_proto = computation->computation().proto();
+  // Output parameter sharding annotations, defaults to REPLICATED(0) if unset.
+  auto tuple_shardings =
+      computation_proto.spmd_output_sharding().tuple_shardings();
+  std::vector output_shardings(tuple_shardings.begin(), tuple_shardings.end());
+  if (output_shardings.empty()) {
+    // Initializes with default sharding type, REPLCIATED.
+    output_shardings.resize(indices.size());
+  }
+  XLA_CHECK(indices.size() == output_shardings.size())
+      << "Expected " << indices.size() << " sharding annotations, but got "
+      << output_shardings.size();
+
+  for (int i = 0; i < indices.size(); ++i) {
+    auto xtensor = (*tensors)[indices[i]];
+    if (output_shardings[i].type()) {
+      // Tensor sharding annotation type is non-zero (sharded).
+      (*sharding_specs)[i] = std::make_shared<XLATensor::ShardingSpec>(
+          output_shardings[i],
+          MakeShapeWithDeviceLayout(
+              xtensor->shape().get(),
+              static_cast<XlaDeviceType>(xtensor->GetDevice().type())));
+      xtensor->SetShardingSpec(*(*sharding_specs)[i]);
+    } else {
+      // Clear sharding if the output parameter is no longer sharded, this
+      // assumes that the output is implicitly replicated and wrapped inside
+      // PjRtShardedData.
+      (*sharding_specs)[i] = std::make_shared<XLATensor::ShardingSpec>(
+          xla::HloSharding::Replicate().ToProto(),
+          MakeShapeWithDeviceLayout(
+              xtensor->shape().get(),
+              static_cast<XlaDeviceType>(xtensor->GetDevice().type())));
+      xtensor->ClearShardingSpec();
+    }
+
+    // Create sharded data placeholder, this will be used to
+    // hold the corresponding computation results for both sharding &
+    // replication.
+    auto sharded_data_placeholder =
+        WrapXlaData(xla::ComputationClient::Get()->WrapDataShards(
+            {}, GetVirtualDevice().toString(),
+            (*sharding_specs)[i]->shape.value(),
+            (*sharding_specs)[i]->sharding));
+
+    // Register the sharded data placeholder to the tensor and its node.
+    (*data_placeholders)[i] = sharded_data_placeholder;
+    xtensor->data()->handle = (*data_placeholders)[i];
+    if (auto ir_value = xtensor->CurrentIrValue()) {
+      if (DeviceData* device_data_node =
+              DeviceData::Cast(ir_value.node.get())) {
+        device_data_node->Assign(xtensor->data()->handle);
```

Comment:
Do we still expect this part of the code being executed? There shouldn't be any IR being generated when there is a placeholder in place. I think we should error out here instead of silently patching the node.

Reply:
This is needed since we don't reset IR when we set placeholders -- https://github.com/pytorch/xla/blob/dc057dcda60355ea38d5686e5baafe677b838564/torch_xla/csrc/xla_graph_executor.cpp#L916

Reply:
I am still a bit confuse in 
https://github.com/pytorch/xla/blob/dc057dcda60355ea38d5686e5baafe677b838564/torch_xla/csrc/xla_graph_executor.cpp#L965-L967
we reset IRs for all of the tensors that will get a new `XlaData` after the execution and in https://github.com/pytorch/xla/blob/dc057dcda60355ea38d5686e5baafe677b838564/torch_xla/csrc/xla_graph_executor.cpp#L915-L923 we assign the placeholder to those XlaTensor(which doesn't have any IR). 

In logic here we are replacing the `unsharded placeholder` with the `sharded_data_placeholder`. I am not sure in what case there will be a `DeviceData` Node that is pointed to the `unsharded placeholder`

Reply:
discussed offline with yeounoh, we will land this pr as it is and I will pick this us and fiugre out what is going on so we don't need to block the benchmark.